source/change-streams/change-streams.rst

```diff
@@ -169,6 +208,12 @@ Driver API
      */
     private resumeToken: Document;
 
```

Comment:
Question for drivers reviewers: How do people feel about adding a public method to retrieve the current resume token? As it stands, users who are implementing their own resume logic would have no way to retrieve the minPromisedResumeToken.

Reply:
Since our guidance for users is to keep track of resume tokens (e.g. https://docs.mongodb.com/ruby-driver/master/tutorials/ruby-driver-change-streams/#resuming-a-change-stream), this seems required to me.

Reply:
If we add this method, would we suggest that users always use this method?

Reply:
👍 in favor of adding a method to get resume token

Reply:
Agreed users should be guided to always use this accessor if its now present.

Reply:
The proposed usage pattern seems a bit cumbersome to me, since it effectively requires the application to consume two parallel "streams" of data from the cursor in order to remain consistent. I'm also quite skeptical that users will follow this advice, or even understand why it's necessary; it's more likely that they'll think "why do I need to call this? I already have the most recent resume token in the document I just read."

Since the `minPromisedResumeToken` is only relevant to the user in cases where the driver cannot resume internally, would it be possible to provide the `minPromisedResumeToken` to the app as part of the exception thrown by the driver on resume failure? Consuming the stream would then look like this:

```
next()
next()
next()
[batch exhausted]
[getMore request fails]
[driver tries to resume with minPromisedResumeToken, fails]
[driver throws exception, bundles minPromisedResumeToken]
[application resumes using minPromisedResumeToken]
```

This seems to me to be a much more natural usage pattern. Our guidance could then be "on failure of a changestream, the resume token returned with the exception is guaranteed to be the most recent point from which it is safe to resume. It is strongly recommended to always use this token."

Reply:
I think there are two use cases for users concerning change stream resume tokens: a) the user wants to resume the stream manually, and b) the user does not want to resume the stream manually.

For the second case, I think we don't need to expose resume tokens in the document, and those users might benefit from being able to remove the resume token all together (as in the case of [SERVER-37669](https://jira.mongodb.org/browse/SERVER-37669).

For the first case, users want to have the best chance of the resume token working when they use it manually. Only being able to retrieve the minPromisedResumeToken on network error might mean that the resume token the user wants to manually restart from is too old to actually use. Here's an example: Say the user has tens of thousands of change streams but changes to each happen less than once before the oplog rolls over. To help ensure that the system administrator can do rolling upgrade, the user takes each resume token and places it in a cache. With the `minPromisedResumeToken` only being accessible via exception, the user cannot retrieve it during normal operation. Therefore, when the system administrator upgrades the application servers, the resume tokens for older change streams won't be valid and will cause the application to create new change streams. If the oplog is large, this means that the server will now have to search through the oplog, even though we already know that there are no entries that will meet the criteria. This may cause performance issue.

For both use cases, the resume token is control information, it allows manipulation of the change stream itself, whereas the rest of the information is the data and metadata about the change. I think it would be advantageous to have a separate path for users to retrieve the control information, and for that path to be consistent.

There is an additional problem with all of the proposed methods, though. In some drivers, if an empty batch is returned from the server, the driver does not return control to the user. This means that even if we do expose a `getResumeToken` method, it will still be difficult for users to retrieve and use that token. I think we need another approach to allowing users to retrieve this information, but that is a driver API change. When we add this new `minPromisedResumeToken`, we should also tell users to not use the resume token in the document and instead retrieve the resume token via that new driver API.

Reply:
resuming with a token explicitly by the application is the only way to resume if the application crashes.  it seems required that there be a way to get real most recent resume time.

Reply:
Seems like this can be resolved?

Reply:
Oh right. TIL you can resolve conversations in GitHub PRs.